pkgs/by-name/ma/material-maker/package.nix

```diff
@@ -0,0 +1,133 @@
+{
+  lib,
+  stdenv,
+  fetchzip,
+  fetchurl,
+  nix-update-script,
+
+  # build deps
+  autoPatchelfHook,
+  copyDesktopItems,
+  makeDesktopItem,
+  undmg,
+
+  # runtime deps
+  libGL,
+  libGLU,
+  libX11,
+  libXcursor,
+  libXext,
+  libXfixes,
+  libXi,
+  libXinerama,
+  libXrandr,
+  libXrender,
+  alsa-lib,
+  libpulseaudio,
+  udev,
+}:
+let
+  pname = "material-maker";
+  version = "1.3";
+  versionUnderscore = "${lib.replaceStrings [ "." ] [ "_" ] version}";
+
+  sources = {
+    linux = fetchzip {
+      url = "https://github.com/RodZill4/material-maker/releases/download/${version}/material_maker_${versionUnderscore}_linux.tar.gz";
+      hash = "sha256-WEu5gVfnswB5zYzu3leOL+hKOBzJbn48gHQKshlfOh4=";
+    };
+    darwin = fetchurl {
+      url = "https://github.com/RodZill4/material-maker/releases/download/${version}/material_maker_${versionUnderscore}.dmg";
+      hash = "sha256-D4jPQEOKws2JFczh0K55mR08jIBMU6jOFOm6TyFxnt8=";
+    };
+  };
+in
+stdenv.mkDerivation (finalAttrs: {
+  inherit pname version;
+
+  src =
+    if stdenv.isLinux && stdenv.isx86_64 then
+      sources.linux
+    else if stdenv.isDarwin then
+      sources.darwin
+    else
+      throw "unsupported platform";
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    copyDesktopItems
+  ] ++ lib.optionals stdenv.isDarwin [ undmg ];
+
+  sourceRoot = ".";
+
+  buildInputs = [
+    libGL
+    libGLU
+    libX11
+    libXcursor
+    libXext
+    libXfixes
+    libXi
+    libXinerama
+    libXrandr
+    libXrender
+  ];
+
+  desktopItems = makeDesktopItem {
+    name = "material-maker";
+    exec = "material-maker";
+    icon = "material-maker";
+    genericName = "Procedural texture generation and 3D model painting tool";
+    desktopName = "Material Maker";
+    comment = finalAttrs.meta.description;
+    terminal = false;
+    categories = [
+      "Graphics"
+      "3DGraphics"
+    ];
+    keywords = [ "material" ];
+  };
+
+  installPhase =
+    ''
+      runHook preInstall
+    ''
+    + lib.optionalString stdenv.isLinux ''
+      install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
+      install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck
+
+      for dir in examples library nodes doc environments meshes export; do
+        cp -R $src/$dir $out/libexec
+      done
+
+      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
+
+      mkdir -p $out/bin
+      ln -s $out/libexec/material-maker $out/bin/material-maker
+    ''
+    + lib.optionalString stdenv.isDarwin ''
+      mkdir -p $out/Applications
+      cp -r ./material_maker.app $out/Applications
+    ''
+    + ''
+      runHook postInstall
+    '';
```

Comment:
```suggestion
    ''
      runHook preInstall
    ''
    + lib.optionalString stdenv.isLinux ''
      install -D -m 755 $src/material_maker.x86_64 $out/opt/material-maker
      install -D -m 644 $src/material_maker.pck $out/opt/material-maker.pck

      mkdir -p $out/share/
      cp -R $src/doc $out/share
    ''
    # Symlink docs if the corresponding output is installed
    # This allows opening documentation directly from the app
    + lib.optionalString (builtins.elem "doc" finalAttrs.meta.outputsToInstall or []) ''
      ln -s $doc/share/doc $out/opt/doc
    ''
    + ''
      for dir in library nodes environments meshes export; do
        cp -R $src/$dir $out/opt
      done

      cp -R $src/examples $examples

      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/

      mkdir -p $out/bin
      ln -s $out/opt/material-maker $out/bin/material-maker
    ''
    + lib.optionalString stdenv.isDarwin ''
      mkdir -p $out/Applications
      cp -r ./material_maker.app $out/Applications
    ''
    + ''
      runHook postInstall
    '';

  outputs = [ "out" "doc" "examples" ];
```

This is a proposal to use `/opt` and to modularize the package into three outputs (the main one, the documentation and the examples), in order to save disk space for a minimal install.

See https://nixos.org/manual/nixpkgs/stable/#chap-multiple-output.

I'm not sure at all that the commented passage is the correct way to have a cross-output symlink, I'd let someone more expert than me comment on this approach.

Reply:
Seems definitely better than using lib exec for sure though shouldn't it all be in the lib optional isLinux bc for Darwin everything is put in the .app/ correct me if I am wrong. Maybe see  the aur for inspiration https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=material-maker-bin

Reply:
Why are you querying meta.inputsToInstall but you've not added it ? and what is the point of the `or []` should add `and stdenv.isLinux` ?

Reply:
> Seems definitely better than using lib exec for sure though shouldn't it all be in the lib optional isLinux bc for Darwin everything is put in the .app/ correct me if I am wrong. Maybe see the aur for inspiration https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=material-maker-bin

Yes, you are right, I forgot `isLinux` in the second block!

> Why are you querying meta.inputsToInstall but you've not added it ? and what is the point of the `or []` should add `and stdenv.isLinux` ?

Also here, I forgot `IsLinux`. 
The `or [ ]` is needed if `outputsToInstall` is not defined (as by default it will install only the "out" output).
Probably you are right and it's better to specify `meta.outputsToInstall = [ "out" ];` and drop `or [ ]`.

(it's my first time tinkering with multiple outputs, feel free to drop the idea if you are not sure it's a good one!)

Reply:
We'll maybe need to also link the examples as it is maybe used by the app, need some testing. I think it's cleaner to declare the outputsToInstall but I'm not sure. I will change some things to add the isLinux and fix the commit and will push this evening.